File: ios/DJMediWallet/DJMediWalletXCode/DJMediWallet25/DJMediWallet25/Core/CredentialManager.swift
//
//  CredentialManager.swift
//  DJMediWallet
//
//  Manages medical credential validation and processing
//

import Foundation

/// Manages validation and processing of medical credentials
public class CredentialManager {
    
    // MARK: - Credential Validation
    
    /// Validate medical credential structure and content
    public func validateCredential(_ credential: MedicalCredential) -> Bool {
        // Check required fields
        guard !credential.id.isEmpty,
              !credential.type.isEmpty,
              credential.issuanceDate <= Date() else {
            return false
        }
        
        // Validate FHIR resource if present
        if let fhirResource = credential.fhirResource {
            return validateFHIRResource(fhirResource)
        }
        
        return true
    }
    
    /// Validate FHIR resource structure
    private func validateFHIRResource(_ resource: FHIRResource) -> Bool {
        // Check resource type
        guard !resource.resourceType.isEmpty else {
            return false
        }
        
        // Validate based on resource type
        switch resource.resourceType {
        case "Patient":
            return validatePatientResource(resource)
        case "Observation":
            return validateObservationResource(resource)
        case "Condition":
            return validateConditionResource(resource)
        case "MedicationStatement":
            return validateMedicationResource(resource)
        case "AllergyIntolerance":
            return validateAllergyResource(resource)
        case "Immunization":
            return validateImmunizationResource(resource)
        default:
            // Allow other resource types
            return true
        }
    }
    
    private func validatePatientResource(_ resource: FHIRResource) -> Bool {
        // Patient should have at least name or identifier
        guard let data = resource.data,
              data["name"] != nil || data["identifier"] != nil else {
            return false
        }
        return true
    }
    
    private func validateObservationResource(_ resource: FHIRResource) -> Bool {
        // Observation must have status and code
        guard let data = resource.data,
              data["status"] != nil,
              data["code"] != nil else {
            return false
        }
        return true
    }
    
    private func validateConditionResource(_ resource: FHIRResource) -> Bool {
        // Condition must have code
        guard let data = resource.data,
              data["code"] != nil else {
            return false
        }
        return true
    }
    
    private func validateMedicationResource(_ resource: FHIRResource) -> Bool {
        // MedicationStatement must have status and medication
        guard let data = resource.data,
              data["status"] != nil,
              (data["medicationCodeableConcept"] != nil || data["medicationReference"] != nil) else {
            return false
        }
        return true
    }
    
    private func validateAllergyResource(_ resource: FHIRResource) -> Bool {
        // AllergyIntolerance must have code or substance
        guard let data = resource.data,
              (data["code"] != nil || data["substance"] != nil) else {
            return false
        }
        return true
    }
    
    private func validateImmunizationResource(_ resource: FHIRResource) -> Bool {
        // Immunization must have status and vaccineCode
        guard let data = resource.data,
              data["status"] != nil,
              data["vaccineCode"] != nil else {
            return false
        }
        return true
    }
    
    // MARK: - SNOMED CT Code Validation
    
    /// Validate SNOMED CT code format
    public func validateSNOMEDCode(_ code: String) -> Bool {
        // SNOMED CT codes are 6-18 digits
        let pattern = "^[0-9]{6,18}$"
        let regex = try? NSRegularExpression(pattern: pattern)
        let range = NSRange(code.startIndex..., in: code)
        return regex?.firstMatch(in: code, range: range) != nil
    }
    
    /// Extract SNOMED codes from FHIR resource
    public func extractSNOMEDCodes(from resource: FHIRResource) -> [SNOMEDCode] {
        var codes: [SNOMEDCode] = []
        
        guard let data = resource.data else {
            return codes
        }
        
        // Recursive function to search for SNOMED codes in nested structures
        func searchForCodes(in value: Any) {
            if let dict = value as? [String: Any] {
                // Check if this is a coding with SNOMED system
                if let system = dict["system"] as? String,
                   system == "http://snomed.info/sct",
                   let code = dict["code"] as? String {
                    let display = dict["display"] as? String
                    codes.append(SNOMEDCode(code: code, display: display))
                }
                
                // Recursively search nested dictionaries
                for (_, nestedValue) in dict {
                    searchForCodes(in: nestedValue)
                }
            } else if let array = value as? [Any] {
                // Recursively search arrays
                for item in array {
                    searchForCodes(in: item)
                }
            }
        }
        
        searchForCodes(in: data)
        return codes
    }
    
    // MARK: - Credential Processing
    
    /// Extract readable summary from credential
    public func extractSummary(from credential: MedicalCredential) -> String {
        guard let resource = credential.fhirResource else {
            return "Medical Credential: \(credential.type)"
        }
        
        switch resource.resourceType {
        case "Patient":
            return extractPatientSummary(resource)
        case "Observation":
            return extractObservationSummary(resource)
        case "Condition":
            return extractConditionSummary(resource)
        case "MedicationStatement":
            return extractMedicationSummary(resource)
        case "AllergyIntolerance":
            return extractAllergySummary(resource)
        case "Immunization":
            return extractImmunizationSummary(resource)
        default:
            return "\(resource.resourceType) Record"
        }
    }
    
    private func extractPatientSummary(_ resource: FHIRResource) -> String {
        guard let data = resource.data else { return "Patient Record" }
        
        if let nameArray = data["name"] as? [[String: Any]],
           let name = nameArray.first,
           let given = name["given"] as? [String],
           let family = name["family"] as? String {
            return "\(given.joined(separator: " ")) \(family)"
        }
        
        return "Patient Record"
    }
    
    private func extractObservationSummary(_ resource: FHIRResource) -> String {
        guard let data = resource.data,
              let code = data["code"] as? [String: Any],
              let coding = code["coding"] as? [[String: Any]],
              let display = coding.first?["display"] as? String else {
            return "Observation"
        }
        return display
    }
    
    private func extractConditionSummary(_ resource: FHIRResource) -> String {
        guard let data = resource.data,
              let code = data["code"] as? [String: Any],
              let coding = code["coding"] as? [[String: Any]],
              let display = coding.first?["display"] as? String else {
            return "Condition"
        }
        return display
    }
    
    private func extractMedicationSummary(_ resource: FHIRResource) -> String {
        guard let data = resource.data else { return "Medication" }
        
        if let medication = data["medicationCodeableConcept"] as? [String: Any],
           let coding = medication["coding"] as? [[String: Any]],
           let display = coding.first?["display"] as? String {
            return display
        }
        
        return "Medication"
    }
    
    private func extractAllergySummary(_ resource: FHIRResource) -> String {
        guard let data = resource.data else { return "Allergy" }
        
        if let code = data["code"] as? [String: Any],
           let coding = code["coding"] as? [[String: Any]],
           let display = coding.first?["display"] as? String {
            return "Allergy: \(display)"
        }
        
        return "Allergy Record"
    }
    
    private func extractImmunizationSummary(_ resource: FHIRResource) -> String {
        guard let data = resource.data,
              let vaccineCode = data["vaccineCode"] as? [String: Any],
              let coding = vaccineCode["coding"] as? [[String: Any]],
              let display = coding.first?["display"] as? String else {
            return "Immunization"
        }
        return "Vaccine: \(display)"
    }
}

// MARK: - Supporting Types

public struct SNOMEDCode: Codable {
    public let code: String
    public let display: String?
    
    public init(code: String, display: String?) {
        self.code = code
        self.display = display
    }
}

File: android/DJMediWallet/app/src/main/java/com/djmediwallet/core/CredentialManager.kt
package com.djmediwallet.core

import com.djmediwallet.models.credential.MedicalCredential
import com.djmediwallet.models.fhir.FHIRResource

/**
 * Manages validation and processing of medical credentials
 */
class CredentialManager {
    
    companion object {
        private const val SNOMED_SYSTEM = "http://snomed.info/sct"
    }
    
    // MARK: - Credential Validation
    
    /**
     * Validate medical credential structure and content
     */
    fun validateCredential(credential: MedicalCredential): Boolean {
        // Check required fields
        if (credential.id.isEmpty() || 
            credential.type.isEmpty() || 
            credential.issuanceDate.after(Date())) {
            return false
        }
        
        // Validate FHIR resource if present
        credential.fhirResource?.let { resource ->
            return validateFHIRResource(resource)
        }
        
        return true
    }
    
    /**
     * Validate FHIR resource structure
     */
    private fun validateFHIRResource(resource: FHIRResource): Boolean {
        if (resource.resourceType.isEmpty()) {
            return false
        }
        
        return when (resource.resourceType) {
            "Patient" -> validatePatientResource(resource)
            "Observation" -> validateObservationResource(resource)
            "Condition" -> validateConditionResource(resource)
            "MedicationStatement" -> validateMedicationResource(resource)
            "AllergyIntolerance" -> validateAllergyResource(resource)
            "Immunization" -> validateImmunizationResource(resource)
            else -> true // Allow other resource types
        }
    }
    
    private fun validatePatientResource(resource: FHIRResource): Boolean {
        val data = resource.data ?: return false
        return data.containsKey("name") || data.containsKey("identifier")
    }
    
    private fun validateObservationResource(resource: FHIRResource): Boolean {
        val data = resource.data ?: return false
        return data.containsKey("status") && data.containsKey("code")
    }
    
    private fun validateConditionResource(resource: FHIRResource): Boolean {
        val data = resource.data ?: return false
        return data.containsKey("code")
    }
    
    private fun validateMedicationResource(resource: FHIRResource): Boolean {
        val data = resource.data ?: return false
        return data.containsKey("status") && 
               (data.containsKey("medicationCodeableConcept") || data.containsKey("medicationReference"))
    }
    
    private fun validateAllergyResource(resource: FHIRResource): Boolean {
        val data = resource.data ?: return false
        return data.containsKey("code") || data.containsKey("substance")
    }
    
    private fun validateImmunizationResource(resource: FHIRResource): Boolean {
        val data = resource.data ?: return false
        return data.containsKey("status") && data.containsKey("vaccineCode")
    }
    
    // MARK: - SNOMED CT Code Validation
    
    /**
     * Validate SNOMED CT code format
     */
    fun validateSNOMEDCode(code: String): Boolean {
        // SNOMED CT codes are 6-18 digits
        return code.matches(Regex("^[0-9]{6,18}$"))
    }
    
    /**
     * Extract SNOMED codes from FHIR resource
     */
    fun extractSNOMEDCodes(resource: FHIRResource): List<SNOMEDCode> {
        val codes = mutableListOf<SNOMEDCode>()
        
        resource.data?.let { data ->
            searchForCodes(data, codes)
        }
        
        return codes
    }
    
    private fun searchForCodes(value: Any, codes: MutableList<SNOMEDCode>) {
        when (value) {
            is Map<*, *> -> {
                // Check if this is a coding with SNOMED system
                val system = value["system"] as? String
                if (system == SNOMED_SYSTEM) {
                    val code = value["code"] as? String
                    val display = value["display"] as? String
                    if (code != null) {
                        codes.add(SNOMEDCode(code, display))
                    }
                }
                
                // Recursively search nested maps
                value.values.forEach { nestedValue ->
                    if (nestedValue != null) {
                        searchForCodes(nestedValue, codes)
                    }
                }
            }
            is List<*> -> {
                // Recursively search lists
                value.forEach { item ->
                    if (item != null) {
                        searchForCodes(item, codes)
                    }
                }
            }
        }
    }
    
    // MARK: - Credential Processing
    
    /**
     * Extract readable summary from credential
     */
    fun extractSummary(credential: MedicalCredential): String {
        val resource = credential.fhirResource ?: return "Medical Credential: ${credential.type}"
        
        return when (resource.resourceType) {
            "Patient" -> extractPatientSummary(resource)
            "Observation" -> extractObservationSummary(resource)
            "Condition" -> extractConditionSummary(resource)
            "MedicationStatement" -> extractMedicationSummary(resource)
            "AllergyIntolerance" -> extractAllergySummary(resource)
            "Immunization" -> extractImmunizationSummary(resource)
            else -> "${resource.resourceType} Record"
        }
    }
    
    @Suppress("UNCHECKED_CAST")
    private fun extractPatientSummary(resource: FHIRResource): String {
        val data = resource.data ?: return "Patient Record"
        
        val nameArray = data["name"] as? List<Map<String, Any>>
        nameArray?.firstOrNull()?.let { name ->
            val given = name["given"] as? List<String>
            val family = name["family"] as? String
            if (given != null && family != null) {
                return "${given.joinToString(" ")} $family"
            }
        }
        
        return "Patient Record"
    }
    
    @Suppress("UNCHECKED_CAST")
    private fun extractObservationSummary(resource: FHIRResource): String {
        val data = resource.data ?: return "Observation"
        
        val code = data["code"] as? Map<String, Any>
        val coding = code?.get("coding") as? List<Map<String, Any>>
        val display = coding?.firstOrNull()?.get("display") as? String
        
        return display ?: "Observation"
    }
    
    @Suppress("UNCHECKED_CAST")
    private fun extractConditionSummary(resource: FHIRResource): String {
        val data = resource.data ?: return "Condition"
        
        val code = data["code"] as? Map<String, Any>
        val coding = code?.get("coding") as? List<Map<String, Any>>
        val display = coding?.firstOrNull()?.get("display") as? String
        
        return display ?: "Condition"
    }
    
    @Suppress("UNCHECKED_CAST")
    private fun extractMedicationSummary(resource: FHIRResource): String {
        val data = resource.data ?: return "Medication"
        
        val medication = data["medicationCodeableConcept"] as? Map<String, Any>
        val coding = medication?.get("coding") as? List<Map<String, Any>>
        val display = coding?.firstOrNull()?.get("display") as? String
        
        return display ?: "Medication"
    }
    
    @Suppress("UNCHECKED_CAST")
    private fun extractAllergySummary(resource: FHIRResource): String {
        val data = resource.data ?: return "Allergy"
        
        val code = data["code"] as? Map<String, Any>
        val coding = code?.get("coding") as? List<Map<String, Any>>
        val display = coding?.firstOrNull()?.get("display") as? String
        
        return display?.let { "Allergy: $it" } ?: "Allergy Record"
    }
    
    @Suppress("UNCHECKED_CAST")
    private fun extractImmunizationSummary(resource: FHIRResource): String {
        val data = resource.data ?: return "Immunization"
        
        val vaccineCode = data["vaccineCode"] as? Map<String, Any>
        val coding = vaccineCode?.get("coding") as? List<Map<String, Any>>
        val display = coding?.firstOrNull()?.get("display") as? String
        
        return display?.let { "Vaccine: $it" } ?: "Immunization"
    }
}

/**
 * SNOMED code data class
 */
data class SNOMEDCode(
    val code: String,
    val display: String?
)

private fun Date(): java.util.Date = java.util.Date()
